File: app/schemas/qa.py
"""
问答模块Schema定义
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


# 会话相关Schema
class ConversationCreate(BaseModel):
    """创建会话请求"""
    title: Optional[str] = Field(None, description="会话标题")
    knowledge_base_ids: Optional[List[str]] = Field(None, description="知识库（文件夹）ID列表")
    top_k: Optional[int] = Field(None, description="检索top K个结果", ge=1, le=200)
    similarity_threshold: Optional[float] = Field(None, description="相似度阈值", ge=0, le=1)
    use_rerank: Optional[bool] = Field(False, description="是否使用重排序")
    rerank_top_n: Optional[int] = Field(None, description="重排序后的top N", ge=1, le=50)


class ConversationUpdate(BaseModel):
    """更新会话请求"""
    title: Optional[str] = Field(None, description="会话标题")
    knowledge_base_ids: Optional[List[str]] = Field(None, description="知识库（文件夹）ID列表")
    top_k: Optional[int] = Field(None, description="检索top K个结果", ge=1, le=200)
    similarity_threshold: Optional[float] = Field(None, description="相似度阈值", ge=0, le=1)
    use_rerank: Optional[bool] = Field(None, description="是否使用重排序")
    rerank_top_n: Optional[int] = Field(None, description="重排序后的top N", ge=1, le=50)


class ConversationResponse(BaseModel):
    """会话响应"""
    id: str
    tenant_id: str
    user_id: str
    title: Optional[str]
    status: str
    config: Optional[Dict[str, Any]]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# 消息相关Schema
class MessageResponse(BaseModel):
    """消息响应"""
    id: str
    conversation_id: str
    role: str
    content: str
    references: Optional[List[Dict[str, Any]]]
    prompt_tokens: Optional[int]
    completion_tokens: Optional[int]
    total_tokens: Optional[int]
    sequence: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class MessageListResponse(BaseModel):
    """消息列表响应"""
    messages: List[MessageResponse]
    total: int


# 对话相关Schema
class ChatRequest(BaseModel):
    """对话请求"""
    query: str = Field(..., description="用户问题", min_length=1)
    knowledge_base_ids: Optional[List[str]] = Field(None, description="知识库（文件夹）ID列表（可选，使用会话配置或默认）")
    top_k: Optional[int] = Field(None, description="检索top K个结果", ge=1, le=200)
    similarity_threshold: Optional[float] = Field(None, description="相似度阈值", ge=0, le=1)
    use_rerank: Optional[bool] = Field(False, description="是否使用重排序")
    rerank_top_n: Optional[int] = Field(None, description="重排序后的top N", ge=1, le=50)


class ChatResponse(BaseModel):
    """对话响应"""
    message_id: str
    content: str
    references: List[Dict[str, Any]]
    usage: Dict[str, Any]

File: frontend/src/api/qa.ts
import client from './client'

export interface Conversation {
  id: string
  tenant_id: string
  user_id: string
  title: string | null
  status: string
  config: Record<string, any> | null
  created_at: string
  updated_at: string
}

export interface Message {
  id: string
  conversation_id: string
  role: string
  content: string
  references: Array<{
    document_id: string
    chunk_index: number
    content: string
    similarity: number
    metadata?: any
  }> | null
  prompt_tokens: number | null
  completion_tokens: number | null
  total_tokens: number | null
  sequence: number
  created_at: string
  updated_at: string
}

export interface ChatRequest {
  query: string
  knowledge_base_ids?: string[]
  top_k?: number
  similarity_threshold?: number
  use_rerank?: boolean
  rerank_top_n?: number
}

export interface ChatResponse {
  message_id: string
  content: string
  references: Array<{
    document_id: string
    chunk_index: number
    content: string
    similarity: number
    metadata?: any
  }>
  usage: {
    prompt_tokens?: number
    completion_tokens?: number
    total_tokens?: number
  }
}

export const qaApi = {
  // 会话管理
  async getConversations(status?: string): Promise<Conversation[]> {
    const params = status ? { status_filter: status } : {}
    const response = await client.get('/qa/conversations', { params })
    return response.data
  },

  async getConversation(conversationId: string): Promise<Conversation> {
    const response = await client.get(`/qa/conversations/${conversationId}`)
    return response.data
  },

  async createConversation(data: {
    title?: string
    knowledge_base_ids?: string[]
    top_k?: number
    similarity_threshold?: number
    use_rerank?: boolean
    rerank_top_n?: number
  }): Promise<Conversation> {
    const response = await client.post('/qa/conversations', data)
    return response.data
  },

  async updateConversation(
    conversationId: string,
    data: {
      title?: string
      knowledge_base_ids?: string[]
      top_k?: number
      similarity_threshold?: number
      use_rerank?: boolean
      rerank_top_n?: number
    }
  ): Promise<Conversation> {
    const response = await client.put(`/qa/conversations/${conversationId}`, data)
    return response.data
  },

  async deleteConversation(conversationId: string): Promise<void> {
    await client.delete(`/qa/conversations/${conversationId}`)
  },

  async archiveConversation(conversationId: string): Promise<Conversation> {
    const response = await client.post(`/qa/conversations/${conversationId}/archive`)
    return response.data
  },

  // 消息管理
  async getMessages(conversationId: string, skip = 0, limit = 100): Promise<{
    messages: Message[]
    total: number
  }> {
    const response = await client.get(`/qa/conversations/${conversationId}/messages`, {
      params: { skip, limit }
    })
    return response.data
  },

  async getMessage(messageId: string): Promise<Message> {
    const response = await client.get(`/qa/messages/${messageId}`)
    return response.data
  },

  async deleteMessage(messageId: string): Promise<void> {
    await client.delete(`/qa/messages/${messageId}`)
  },

  // 对话
  async chat(conversationId: string, data: ChatRequest): Promise<ChatResponse> {
    const response = await client.post(`/qa/conversations/${conversationId}/chat`, data)
    return response.data
  },

  async chatStream(
    conversationId: string,
    data: ChatRequest,
    onChunk: (chunk: string) => void,
    onComplete?: () => void,
    onError?: (error: Error) => void
  ): Promise<void> {
    try {
      const token = localStorage.getItem('token')
      const baseURL = client.defaults.baseURL || '/api/v1'
      const url = `${baseURL}/qa/conversations/${conversationId}/chat/stream`
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': token ? `Bearer ${token}` : ''
        },
        body: JSON.stringify(data)
      })

      if (!response.ok) {
        const errorText = await response.text()
        throw new Error(`HTTP error! status: ${response.status}, ${errorText}`)
      }

      const reader = response.body?.getReader()
      const decoder = new TextDecoder()

      if (!reader) {
        throw new Error('Response body is not readable')
      }

      let buffer = ''
      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n\n')
        
        // 保留最后一个不完整的行
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (line.trim().startsWith('data: ')) {
            const dataStr = line.trim().slice(6)
            if (dataStr === '[DONE]') {
              onComplete?.()
              return
            }
            if (dataStr) {
              onChunk(dataStr)
            }
          }
        }
      }
      
      // 处理剩余的buffer
      if (buffer.trim().startsWith('data: ')) {
        const dataStr = buffer.trim().slice(6)
        if (dataStr && dataStr !== '[DONE]') {
          onChunk(dataStr)
        }
      }
      
      onComplete?.()
    } catch (error) {
      onError?.(error as Error)
    }
  }
}
